### PR TITLE
[Deprecated] - support eos multisig

### PR DESCRIPTION
**Pull request by bas4r**

### CODE REVIEW

**Comment by deckb** [comment]
@basarrcan I do not think you committed the model containing `EosAuthorizationPerm`. When I build I get the message: 

```
src/chains/eos_2/eosTransaction.ts:17:3 - error TS2724: '"./models"' has no exported member named 'EosAuthorizationPerm'. Did you mean 'EosAuthorizationStruct'?

17   EosAuthorizationPerm,
```

File: src/chains/eos_2/eosTransaction.ts [hunk 1]
```diff
@@ -327,12 +335,21 @@ export class EosTransaction implements Transaction {
     return !isNullOrEmpty(this.signatures)
   }
 
+  private checkAuthSigned(auth: EosRequiredAuthorization): boolean {
+    const weights: number[] = []
+    auth.keys.forEach(keyObj => weights.push(this.hasSignatureForPublicKey(keyObj.key) ? keyObj.weight : 0))
```
Comment (deckb) [inline src/chains/eos_2/eosTransaction.ts:340]: I'm getting an error when building:
```
src/chains/eos_2/eosTransaction.ts:368:47 - error TS7006: Parameter 'keyObj' implicitly has an 'any' type.

368       auth.requiredAuthorization.keys.forEach(keyObj => {
                                                  ~~~~~~
```

**Review by deckb — CHANGES_REQUESTED** [review]
Please commit the missing file.

**Comment by basarrcan** [comment]
> Please commit the missing file.

committed

File: src/chains/eos_2/eosTransaction.ts [hunk 2]
```diff
@@ -345,9 +362,13 @@ export class EosTransaction implements Transaction {
 
   /** An array of authorizations (e.g. account/permission) that do not have an attached signature
    *  Retuns null if no signatures are missing */
-  public get missingSignatures(): EosAuthorization[] {
+  public get missingSignatures(): EosAuthorizationPerm[] {
```
Comment (deckb) [inline src/chains/eos_2/eosTransaction.ts:365]: This getter function doesn't work as expected anymore.

I believe we actually want to confirm that the threshold is met for each action. Something like:
```
const missing = this._requiredAuthorizations?.filter(auth => {
      return !this.checkAuthSigned(auth.requiredAuthorization);
    })
```

**Comment by WarrickFitz** [comment]
I'm going to wait until your conversation thread is resolved before I review @basarrcan @deckb

File: src/chains/eos_2/eosTransaction.ts (hunk 2 above)
Comment (basarrcan) [reply src/chains/eos_2/eosTransaction.ts:365]: done

File: src/chains/eos_2/eosTransaction.ts [hunk 3]
```diff
@@ -27,14 +29,14 @@ import { throwAndLogError, throwNewError } from '../../errors'
 import { ChainSettingsCommunicationSettings, ConfirmType } from '../../models'
 import { Transaction } from '../../interfaces'
 
-export type PublicKeyMapCache = {
+export type PermissionMapCache = {
```
[the remark this answers is not in the document]
Comment (basarrcan) [reply src/chains/eos_2/eosTransaction.ts]: moved to generalModels since we have other auth related types there.

File: src/chains/eos_2/eosTransaction.ts [hunk 4]
```diff
@@ -381,7 +397,11 @@ export class EosTransaction implements Transaction {
 
   /** TODO: Implement support for eos multi-sig transactions */
```
Comment (WarrickFitz) [inline src/chains/eos_2/eosTransaction.ts:398]: This comment can be removed

**Review by WarrickFitz — APPROVED** [review]
Beside the comment that should be removed this seems to work as expected.

File: src/chains/eos_2/eosTransaction.ts (hunk 4 above)
Comment (basarrcan) [reply src/chains/eos_2/eosTransaction.ts]: removed

File: src/chains/eos_2/eosCreateAccount.ts [hunk 5]
```diff
@@ -93,6 +102,48 @@ export class EosCreateAccount implements CreateAccount {
...
+      this._authorizations.active = isValidEosPublicKey(multisigOptions?.active as EosPublicKey)
+        ? publicKeyToAuth(multisigOptions.active as EosPublicKey)
+        : msigOptionToAuth(multisigOptions.active as MsigAuthOptions)
+    }
+  }
+
+  /** Only available if NOT multisig */
+  get publicKeys(): EosPublicKeys {
+    if (this.isMultisig) {
+      throwNewError('get publicKeys can NOT be called if multisigOptions is provided.')
```
Comment (deckb) [inline src/chains/eos_2/eosCreateAccount.ts:139]: Why is this the case? Couldn't this just return an array of keys to owner and active? 
Something like:
```
     owner: this._authorizations?.owner?.keys,
      active: this._authorizations?.active?.keys,
```
I might be missing the purpose of this function but it seems we also really want to return all permissions not just owner and active

**Review by deckb — COMMENTED** [review]
It seems the creating and signing of multisig are at odds with each other. It looks like we only support creating an account with existing accounts and signing only supports signing with just keys. Am I missing something?

File: src/chains/eos_2/eosCreateAccount.ts (hunk 5 above)
Comment (basarrcan) [reply src/chains/eos_2/eosCreateAccount.ts:139]: > It seems the creating and signing of multisig are at odds with each other. It looks like we only support creating an account with existing accounts and signing only supports signing with just keys. Am I missing something?

creating account supports both plain publicKeys and accounts(with permission name). Signing of course requires private key regardless if multisig account permissions consist of only publicKey or account/permission tuple.

Comment (basarrcan) [reply src/chains/eos_2/eosCreateAccount.ts:139]: > Why is this the case? Couldn't this just return an array of keys to owner and active? Something like:
> 
> ```
>      owner: this._authorizations?.owner?.keys,
>       active: this._authorizations?.active?.keys,
> ```
> 
> I might be missing the purpose of this function but it seems we also really want to return all permissions not just owner and active

Now instead of passing publicKeyOwner and publicKeyActive into native eos account creation, we convert publicKeys or multisigOptions info provided from transaction options into this._authorization and pass ownerAuth and activeAuth into create native account template. get publicKeys() getter is created for using with virtual or createescrow account creations which only supports plain publicKey for active and owner permissions. Its also used with generateKeysIfNeeded, which is only needs to be called if account is not multisig. If multisigOptions passed account creation has to be native and we require all permissions specified, thus there is no use for get publicKeys.

Comment (deckb) [reply src/chains/eos_2/eosCreateAccount.ts:139]: > creating account supports both plain publicKeys and accounts(with permission name). Signing of course requires private key regardless if multisig account permissions consist of only publicKey or account/permission tuple.

I think I misspoke so let me try again. 

I do not believe the example that you give for creating msig accounts will not work with the way msig signing has been implemented see `hasSignatureForAuthorization`. Only the `keys` portion is checked because checking for accounts is a whole other problem. 

Basically we should just show creating an account with keys and not accounts since chainjs cannot currently do anything with a msig account created with accounts.

Comment (deckb) [reply src/chains/eos_2/eosCreateAccount.ts:139]: We could do that but the reason I didn't go down that rabbit hole is because you could have n number of accounts. For example:

accountA:
   active: accountB@active
   
accountB:
   active: accountC@active

accountC:
   active: accountD@active
   
   
 So as far as I can see a deeper discussion about what we want to support is required because in theory the nesting can go on for a while and that could cause performance issues as well as just tons of calls to apis to resolve keys. Plus we haven't even talked about what if the permission is itself a msig? 
 
 In my head the first pass should just show keys and we can design a good solution for supporting accounts.

Comment (deckb) [reply src/chains/eos_2/eosCreateAccount.ts:139]: @traylewin thoughts?

Comment (basarrcan) [reply src/chains/eos_2/eosCreateAccount.ts]: I think supporting that and using it on oreVault provides great value, giving allowing jointAccount owners to take ownership of their chainAccount and still be one of the owner of the joint account. Since the accountName is the real address on EOS paradigm, it would also be consistent with other chains in my opinion. 

As far as the nested permission issue goes, its just possible to overwhelm apis in other ways like just appending a lot of permissions to a single action, or trying to sign a transaction with a lot of actions in it. So to me, not supporting it doesn't add any benefit. 

Also to note, this is not just a multisig issue. Currently chain-js as a generic library does not work with accounts that have accountName/permissionName attached to its permissions. Hardcoded to expect keys array.

Comment (basarrcan) [reply src/chains/eos_2/eosCreateAccount.ts]: @deckb I updated eosTransaction class to support accounts[] field IF the provided accountName/permission tuple for authorization does not contain another accounts[] for its authorization (hence nested accounts[] auth is not supported for now, and throws error). I think this provides the support without getting things too complicated and prone to execution errors. And it will allow us to use oreVault the way we initially decided with @traylewin

**Review by deckb — APPROVED** [review]
Some minor comment tweaks but looks ok

**Comment by deckb** [comment]
@WarrickFitz requested a review again in case you wanted to take a look at the account creation and support of accounts in msigs.

**Comment by WarrickFitz** [comment]
I'm cleaning up the PR backlog. This has been open for a while and these features are in the product now so we must have merged these changes another way.
